src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1000,6 +1026,33 @@ private Task WriteToStreamAsync(ReadOnlyMemory<byte> source)
             return _stream.WriteAsync(source);
         }
 
+        private bool TryReadNextLine(out ReadOnlySpan<byte> line)
```

Comment:
It would be nice to not duplicate the logic between here and ReadNextLineAsync.

That said, I assume the hard part is dealing with Span, right?

Reply:
Yeah.  I can use the Try from the async one if I change it to use ArraySegment instead of Span.  I'll do that, but 😦

Reply:
yeah, not ideal :(

ArraySegment or Memory?

Reply:
It's also annoying because the async one needs to track how much it's already scanned, which the synchronous caller doesn't need.

Reply:
ArraySegment, simply so I don't then have to go update all of the existing callers to ReadNextLineAsync and don't have to convert between them.  If we want to clean that up subsequently, we can.

Reply:
Ok.

In the future, for perf, it may make sense for callers of the existing ReadLineAsync to do TryReadLine instead (and only do the async call if it fails) to avoid the overhead of the async call.

This gets tricky with the logic to not rescan stuff we've already rescanned, but maybe that's less important.

Just food for thought.